src/vmm/src/device_manager/persist.rs

```diff
@@ -277,8 +277,23 @@ impl<'a> Persist<'a> for MMIODeviceManager {
                     .map_err(Error::Legacy)?;
 
                     dev_manager
-                        .register_mmio_serial(vm, serial, Some(state.mmio_slot.clone()))
+                        .register_mmio_serial(vm, serial.clone(), Some(state.mmio_slot.clone()))
                         .map_err(Error::DeviceManager)?;
+
+                    // When restoring from a previously saved state, there is no serial
+                    // driver initialization, therefore the RDA (Received Data Available)
+                    // interrupt is not enabled. Because of that, the driver won't get
+                    // notified of any bytes that we send to the guest. The clean solution
+                    // would be to save the whole serial device state when we do the vm
+                    // serialization. For now we set that bit manually
+                    serial
+                        .lock()
+                        .expect("Poisoned lock")
+                        .serial
+                        .write(1, crate::IER_RDA_BIT)
+                        .map_err(|_| {
+                            Error::Legacy(crate::Error::Serial(std::io::Error::last_os_error()))
+                        })?;
```

Comment:
This can be removed from here and added to the specialized function you created (`serial_set_rda()`).
```suggestion
```